src/classes/OPP_OpportunityNaming.cls

```diff
@@ -0,0 +1,392 @@
+            splitField.remove(splitField.size()-1);
+
+            //use the correct sObject
+            sObject thisObj;
+            if (!splitField.isEmpty() && splitField[0].equalsIgnoreCase('contact')) {
+                thisObj = con;
+                splitField.remove(0);
+            } else {
+                thisObj = opp;
+            }
+
+            //traverse parent relationships until the last one
+            if (thisObj!=null) {
+                for (string parentObj : splitField) {
+                    thisObj = thisObj.getsObject(parentObj);
+                }
+            }
+
+            //if this field is a date or datetime field, follow the format supplied in the settings
+            string val = getFieldValue(thisObj, fieldName, ons);
+
+            // add back the {}'s for string substitution
+            strField = chLToken + strField + chRToken;
+            if (val == null) {   
+                // eat potential trailing space
+                result = result.replace(strField + ' ', '');
+                // eat potential leading space
+                result = result.replace(' ' + strField, '');
+            }
+            result = result.replace(strField, val == null ? '' : val);
+        }
+        return result;
+    }
+
+     /*******************************************************************************************************
+    * @description Gets the field value from an object, and formats it properly if needed.
+    * @param obj The sObject to get the field value from.
+    * @param fieldName The name of the field to get.
+    * @param ons The opportunity naming setting that applies to this field.
+    * @return string The formatted value of the field.
+    ********************************************************************************************************/
+   public static string getFieldValue(sObject obj, string fieldName, Opportunity_Naming_Settings__c ons) {
+        if (obj==null) return null;
+        Schema.DisplayType dt = UTIL_Describe.getFieldDisplaytype(obj.getSObjectType().getDescribe().getName(),fieldName);
+        string result = '';
+
+        //If this is a Date or DateTime field, use the specified formatting if there is one
+        if (dt == Schema.DisplayType.Date || dt == Schema.DisplayType.DateTime) {
+            DateTime d = (DateTime) obj.get(fieldName.trim());
+            string dateFormat = String.isBlank(ons.Date_Format__c) ? 'yyyy.MM.dd' : ons.Date_Format__c;
+            
+            //if this was a date, use the GMT time zone so we don't end up on a different date
+            if (dt == Schema.DisplayType.Date) {
+                result = d.formatGmt(dateFormat);
+            } else {
+                result = d.format(dateFormat);
+            }
+        } else if (dt == Schema.DisplayType.Currency) {
+            result = Label.npe01.DefaultContactTransactionCurrency + String.valueOf(obj.get(fieldName.trim()));
```

Comment:
isn't this currency field coming from the Opp, and thus we should use the currency of the opp?